src/main/scala/formal/FormalUtils.scala

```diff
@@ -0,0 +1,247 @@
+      case _ => ""
+    }
+    val proposed_src = prop_type.toString + "_" + line_info
+
+//    var max_rand = 15
+//    val random_src = scala.util.Random
+//    while (!prop_name_set.add(proposed_src)) {
+//      proposed_src = src + "_" + "%x".format(random_src.nextInt(max_rand))
+//      max_rand = max_rand * 2
+//    }
+
+    val src_wrap = s"@[${proposed_src}]"
+    if (dir==MonitorDirection.Monitor) {
+//      assert(cond, s"Assert: ${prop_type.toString} ${message} ${src_wrap}")
+      when(!cond) {
+        printf(s"assert:${proposed_src}:${prop_type.toString} ${message}")
+      }
+    } else if (dir==MonitorDirection.Receiver) {
+//      assert(cond, s"Assert: ${prop_type.toString} ${message} ${src_wrap}")
+      when(!cond) {
+        printf(s"assert:${proposed_src}:${prop_type.toString} ${message}")
+      }
+    } else if (dir==MonitorDirection.Driver) {
+//      assert(cond, s"Assume: ${prop_type.toString} ${message} ${src_wrap}")
+      when(!cond) {
+        printf(s"assume:${proposed_src}:${prop_type.toString} ${message}")
+      }
+    } else if (dir==MonitorDirection.Cover) {
+        if (prop_type==PropertyClass.CoverDisableMonitor) {
+//          assert(!cond, s"Assert: ${prop_type.toString} ${message} ${src_wrap}")
+          when(cond) { //We want to assert that the condition is never true, which is opposite of a normal assertion
+            printf(s"assert:${proposed_src}:${prop_type.toString} ${message}")
+          }
+        } else {
+//        chisel3.core.withReset(false.B) {
+//          assert(!cond, s"Cover: ${prop_type.toString} ${message} ${src_wrap}")
+            when(cond) {
+              printf(s"cover:${proposed_src}:${prop_type.toString} ${message}")
+            }
+//        }
+        }
+    }
+  }
+
+  def apply(dir: MonitorDirection, cond: Seq[Seq[Bool]], crosscond: Bool, message: String,
+            prop_type: PropertyClass,
+            idx: String)(implicit sourceInfo: SourceInfo): Unit = {
+    if (cond.isEmpty) {
+      Property(dir, crosscond, message, prop_type, idx)
+    } else {
+      cond.head.zipWithIndex.foreach( {case (crossval, cnt) =>
+        Property(dir, cond.tail, crossval && crosscond, message, prop_type, idx + "_" + cnt)
+      })
+    }
+  }
+  def apply(dir: MonitorDirection, cond: Seq[Seq[Bool]], crosscond: Bool,
+            message: String)(implicit sourceInfo: SourceInfo): Unit = {
+    Property(dir, cond, crosscond, message, PropertyClass.Default, "")
+  }
+  def apply(dir: MonitorDirection, cond: Seq[Seq[Bool]], crosscond: Bool, message: String,
+            prop_type: PropertyClass)(implicit sourceInfo: SourceInfo): Unit = {
+    Property(dir, cond, crosscond, message, prop_type, "")
+  }
+  def apply(cond: Bool)(implicit sourceInfo: SourceInfo): Unit ={
+    Property(MonitorDirection.Monitor, cond, "Sanity Property", PropertyClass.LocalRTL, "")
+  }
+}
+
+object SourceGet {
+  def get_line_num(implicit sourceInfo: SourceInfo): String = {
```

Comment:
camelCase is the preferred style for scala methods (as opposed to chisel wires).

Reply:
I'm moving a file from federation to rocket chip, this cannot be changed without changing the corresponding calls (in federation).